teps/0089-spire-support.md

```diff
@@ -0,0 +1,262 @@
+3. Tekton Pipelines can't verify that the results it reads weren't modified
+
+This is where SPIRE comes in!
+SPIRE can be used to request signatures and certificates (SVIDs) for a given workload.
+This workload can be anything, including a Result, a Pod spec, or an entire TaskRun yaml.
+
+We're going to use SPIRE to mitigate all three of the issues mentioned above.
+
+## SPIRE Concepts
+Before discussing how SPIRE is going to fix these issues, here's a very basic overview of how it works:
+
+Pods running in the cluster can interact with SPIRE.
+A pod will send some `payload` to SPIRE for signing.
+The SPIRE socket will confirm that the request came from an approved workload (in our case, the Tekton controller or Pods created by the controller) and then it will sign the `payload`.
+SPIRE will return:
+1. The `signature` over the `payload`
+1. A SPIFFE Verifiable Identity Document, or [SVID](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid), an X509 certificate containing the public key used to verify the `signature` against the `payload`
+
+SPIRE can be used to request certificates (SVIDs) for a given k8s workload and can use these certificates to create signatures for the data we need to verify, i.e. an entire TaskRun yaml and results. For our use case, both the Tekton entrypointer image (running in Pods) and the Tekton Controller will interact with SPIRE.
+
+### Installing SPIRE
+SPIRE runs as a Unix domain socket on the k8s node.
+We can use [spiffe-csi](https://github.com/spiffe/spiffe-csi) to mount the SPIRE socket into Pods as a `csi` type Volume so that we don't have to rely on the `hostPath` volume.
+Users will be responsible for installing this themselves.
+When creating Pods, we would automatically mount this volume in as appropriate.
+
+This volume mount would look something like this on an arbitrary Pod created by the controller:
+
+```yaml
+containers:
+- name: my-image
+    volumeMounts:
+    - name: spiffe-workload-api
+      mountPath: /spiffe-workload-api
+      readOnly: true
+    env:
+    - name: SPIFFE_ENDPOINT_SOCKET
+      value: unix:///spiffe-workload-api/spire-agent.sock
+  volumes:
+  - name: spiffe-workload-api
+    csi:
+      driver: "csi.spiffe.io"
+```
+
+
+## Proposed Solution
+
+### 1. Tekton Chains has no way to verify that a TaskRun in the cluster was created by Tekton
+The solution to this relies on the SPIRE [k8s Workload Attestor](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md).
+SPIRE can be configured to only issue SVIDs to certain k8s workloads.
+That means we can set up SPIRE to only issue signatures for:
+* The Tekton controller
+* Pods created by the Tekton controller (e.g. pods running the entrypointer image)
+
+We can be certain that all signed Tasks were created by Tekton itself, since no other Pods can request signatures.
```

Comment:
couple nits:
- 'all signed Tasks' -> 'all signed TaskRuns' (signing of the Tasks themselves would be a different story)
- 'were created by Tekton' -> 'were only updated by Tekton once execution started' or similar (see comment above)